Yosemite/Yosemite/Actions/CouponAction.swift

```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+/// Defines the `actions` supported by the `CouponStore`.
+///
+public enum CouponAction: Action {
+
+    /// Retrieves and stores Coupons for a site
+    ///
+    /// - `siteID`: the site for which coupons should be fetched.
+    /// - `pageNumber`: page of results based on the `pageSize` provided. 1-indexed.
+    /// - `pageSize`: number of results per page.
+    /// - `onCompletion`: invoked when the sync operation finishes.
+    ///     - `result.success(Bool)`: value indicates whether there are further pages to retrieve.
+    ///     - `result.failure(Error)`: error indicates issues syncing the specified page.
+    ///
+    case synchronizeCoupons(siteID: Int64,
+                            pageNumber: Int,
+                            pageSize: Int,
+                            onCompletion: (Result<Bool, Error>) -> Void)
```

Comment:
What do you think about having a tuple of a list of Coupons and a boolean that indicates whether there are further pages to retrieve? 
Something like that `onCompletion: (Result<(coupons: [Coupon], hasNextPage: Bool), Error>) -> Void)`

Reply:
I took cues from the existing `synchronize` actions in designing this one, and found that none of the paginated actions include the model objects in the `onCompletion` arguments. There are a few non-paginated actions which do, and actions for some paginated endpoints which have a `retrieve` method that may return the models, e.g. on `[ProductAction](https://github.com/woocommerce/woocommerce-ios/blob/e8f9a518a50c1d7200ce86c6d6122320da658335/Yosemite/Yosemite/Actions/ProductAction.swift#L37)`.

If we passed in the coupons, would you expect the completion handler to receive only the coupons from the page requested, or all coupons retrieved so far? I would think just the ones from the page specified, personally, but that list seems like it would be of limited use to the view.

I'm not sure that this would be the right way to go for the `synchronizeCoupons` action, as returning the results seems like a side effect. If there are future uses of the store which may benefit from getting the results back, I think it might be better to add a `retrieveCoupons` action, which provides the model objects to the completion handler. That said, perhaps we'd always want to store coupons we'd gone to the trouble of fetching from the API, so synchronize would still be the right method to call in that situation too. 

My understanding is that accessing model objects for display is usually a two part process: firstly fetch and store the model (synchronize) and secondly use a `FetchedResultsController` to retrieve them from storage for display in a table view. Perhaps, for other use cases which require the results in response to an action instead of via a `FetchedResultsController`, we could also provide Yosemite methods for getting models from Storage, which could be used within completion handlers to compose the required behaviour that way. Have I understood the approach correctly here, wrt `FetchedResultsController`s?

For the coupon management screen, I don't think we need the completion handler to receive the array of coupons, because its `FetchedResultsController` will handle getting the models from Core Data, but I'm wary that I might be missing the point that you're making with this suggestion, so I'm definitely happy to discuss further.

Reply:
> For the coupon management screen, I don't think we need the completion handler to receive the array of coupons, because its FetchedResultsController will handle getting the models from Core Data, but I'm wary that I might be missing the point that you're making with this suggestion, so I'm definitely happy to discuss further.
 

It's totally ok to left it as you implemented it, it makes totally sense to use the `FetchedResultsController`, also because, maybe, we want to show cached coupons before fetching the new ones. 💯